### ISSUE TITLE
V14-RC2: Login when hosting behind proxy

### ISSUE THREAD

**Issue by enkelmedia** [umbraco/Umbraco-CMS#16179]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

v14-RC2

### Bug summary

I'm not sure this is a bug, it might also just be a configuration thing. I'm trying to run v14 using Visual Studio "Dev Tunnels" which is basically a reverse proxy and getting this errors.

### Specifics

I'm getting this error:

```
error:invalid_request
error_description:The specified 'redirect_uri' is not valid for this client application.
error_uri:https://documentation.openiddict.com/errors/ID2043
```

![umb-dev-tunnel](https://github.com/umbraco/Umbraco-CMS/assets/1782524/306be98d-c40f-4778-b029-8228adfc19b7)

I haven't looked at the code but I did try to configure this domain using "Culture and Hostnames" which did not seem to work.

### Steps to reproduce

* Create new Umbraco instance
* Setup "Dev Tunnel" in Visual Studio
* Run application, to to `/umbraco`

### Expected result / actual result

I'm not sure about how the internals work, the default should probably be to allow login from any configured domain (and maybe `localhost` and `*.local*.

In this particular scenario the exact URL to the dev tunnel might change over time so it would be nice to be able to allow wildcards `*.euw.devtunnels.ms`.

**Comment by bergmania** [comment]
Sounds like this is related to the `BackOfficeApplicationManager.EnsureBackOfficeApplicationAsync()`, where OpenIddict need to know where the backoffice is located.

On top of my head, i dont have a good solution to how this can be resolved when using reverse proxy.

**Comment by enkelmedia** [comment]
As far as I understand this any redirect URL has to be whitelisted when registering the OpenId Connect app with OpenDict.

Somewhere around here:

https://github.com/umbraco/Umbraco-CMS/blob/d4188022af2a0c7d8ebc6214ff4ed1d29f7865ed/src/Umbraco.Cms.Api.Management/Security/BackOfficeApplicationManager.cs#L109

Another scenario, that is supported in v13, that might be impacted here is if you host the same instance but on different hostnames (e.g. mysite.se, mysite.dk, mysite.com) and would like to allow access to the backoffice from mysite.se/umbraco, mysite.dk/umbraco and mysite.com/umbraco. In that case, a single backoffice URL is not enough.

Also, the default "thing" in Umbraco that figures out the current application URL could maybe take the `X-Forward-For` into consideration, that could probably solve the problem of running behind a proxy.

But I came to think about this, I solved this in v13 with a hack:

```
public class ProxyHostRewriteRule : Microsoft.AspNetCore.Rewrite.IRule
{
    public void ApplyRule(RewriteContext context)
    {
        context.Result = RuleResult.ContinueRules;

        // Get a reference to the current request
        HttpRequest request = context.HttpContext.Request;

        // Get the host from the header
        string? proxyHost = request.Headers["x-forwarded-host"];
        if (!string.IsNullOrWhiteSpace(proxyHost))
        {
            // Update the host
            request.Host = new HostString(proxyHost);
        }

        string? proxyScheme = request.Headers["x-forwarded-scheme"];
        if (!string.IsNullOrWhiteSpace(proxyScheme))
        {
            // Update the scheme (http/https)
            request.Scheme = proxyScheme;
        }

    }
}
```

Then during startup:

```
var rewriteOptions = new RewriteOptions();
rewriteOptions.Rules.Add(new ProxyHostRewriteRule());
app.UseRewriter(rewriteOptions);
```

I need to verify if this would solve the problem on v14, I'll keep you posted.

**Comment by LennardF1989** [comment]
Heya! Got the same issue on the latest RC! We started out on https://localhost:8080 and have since moved to http://localhost:8080 and https://localhost:8443 - but when going to Umbraco, it will not accept the redirect_uri it creates (which points at https://localhost:8443/umbraco).

Most likely, this is due to the registration of the application when its first created, but we probably need a way to easily override this. Will investigate further.

**Comment by tobias-johansson-nltg** [comment]
I think we have the same issue.

We are logging our users in using their Microsoft accounts, that are managed from Azure (Entra Ids).
This worked fine in Umbraco 13, using the setup for OpenIdConnect described [here](https://docs.umbraco.com/umbraco-cms/reference/security/external-login-providers).
When upgrading to Umbraco 14 this breaks, while still using the same `Microsoft.AspNetCore.Authentication.OpenIdConnect` version `8.0.6`. This error occurrs when logging in:

![image](https://github.com/umbraco/Umbraco-CMS/assets/154510886/640cbd28-31e9-4aa5-8477-44853102ae99)

So I tried to use `Microsoft.AspNetCore.Authentication.MicrosoftAccount` instead. It didn't get the `iss` error, but we instead got the error described in this issue:

![image](https://github.com/umbraco/Umbraco-CMS/assets/154510886/7161be7f-eb83-4b21-9387-6442162b546d)

Our setup is as follows:
We run Umbraco in Azure, with an Azure App Service. Let's say the address to the service is `https://our-umbraco-instance.azurewebsites.net`. The traffic goes in through Azure Front Door, which our external domain is connected to. Let's call the domain `https://our-umbraco-instance.com`. So the traffic goes like this:

User -> AFD (`https://our-umbraco-instance.com`) -> AAS (`https://our-umbraco-instance.azurewebsites.net`)

I have tried to understand where things go wrong, and these are my findings:
Running locally via `https://localhost:<port>` --> __Works__
Running locally with a local nginx proxy via `https://my-machine` --> __Same error as in this issue__
Accessing via AAS (`https://our-umbraco-instance.azurewebsites.net`) --> __Works__
Accessing via AFD (`https://our-umbraco-instance.com`) --> __Same error as in this issue__

As I said, this worked in Umbraco 13. What has changed, and what can we do to resolve the issue?

**Comment by H-WalkerNltg** [comment]
Hi! Could you please provide any updates regarding this issue? Dennis from support suggested I reach out here for updates on this urgent issue. Thank you!

**Comment by AndersBrohus** [comment]
We have this happens on 14.0.0 - While hosting on a "normal" IIS setup :)

**Comment by bergmania** [comment]
Should be fixed by https://github.com/umbraco/Umbraco-CMS/pull/16614

**Comment by tobias-johansson-nltg** [comment]
We get the same error after upgrading to 14.1.0-rc

**Comment by bergmania** [comment]
@tobias-johansson-nltg thanks for reporting, Ill take a look

**Comment by nikolajlauridsen** [comment]
Fixed in #16660 😄

**Comment by bergmania** [comment]
To use a proxy, it will be required to set the ApplicationUrl in config. The reason is, OpenIddict need to know where the requests will be coming from. The fix will be included in the 14.1 final or next RC, if we create one

**Comment by tobias-johansson-nltg** [comment]
> To use a proxy, it will be required to set the ApplicationUrl in config. The reason is, OpenIddict need to know where the requests will be coming from. The fix will be included in the 14.1 final or next RC, if we create one

Thanks!
Will you have updated documentation of how to set the URL in config?

**Comment by bergmania** [comment]
> > To use a proxy, it will be required to set the ApplicationUrl in config. The reason is, OpenIddict need to know where the requests will be coming from. The fix will be included in the 14.1 final or next RC, if we create one
> 
> Thanks! Will you have updated documentation of how to set the URL in config?

We are reusing the existing config, that we recommend using in production anyway https://docs.umbraco.com/umbraco-cms/extending/health-check/guides/fixedapplicationurl

**Comment by enkelmedia** [comment]
Just adding to this, @bergmania - what if someone needs to access the backoffice from multiple different domains?

I think it would make sense to allow all configured domains in “Culture and hostnames”? Or at least provide some kind of extension point when the value is provided so that people can provide their own list of allowed redirects.

**Comment by david-tattan** [comment]
Hi, we are still having issues with the login screen and reverse proxies. We are using Azure Front Door, with a custom domain eg. https://my.custom.domain/

We can login successfully, but:
https://my.custom.domain/umbraco/management/api/v1/security/back-office/authorize?redirect_uri=......
returns a 302, and redirects to the default domain of the Azure Service App that is behind the Front Door:
![image](https://github.com/user-attachments/assets/ad7dfb8c-f733-48d2-9a0c-7e50acb40f37)

On successful authentication, we are redirected back to https://my.custom.domain, so it is only on the openid authorization step that we see this issue right now.

We have added these configurations, with the values set to the custom domain of the Front Door:

```
Umbraco__CMS__Security__BackOfficeHost
Umbraco__CMS__WebRouting__UmbracoApplicationUrl
```

We've also tried adding an applicationHost.xdt file to allow HTTP_X_ORIGINAL_HOST and HTTP_HOST, but that didn't seem to change anything
